metricsaggregator/src/main/java/io/dockstore/metricsaggregator/client/cli/TerraMetricsSubmitter.java

```diff
@@ -0,0 +1,572 @@
+            return Optional.empty();
+        } else if (foundFullWorkflowPaths.size() > 1) {
+            // There is already a workflow in the same repository with the same descriptor path that we're looking for.
+            // Skip this source_url because it is an ambiguous case and we can't identify which workflow the source url is referring to.
+            /*
+            logSkippedExecution(sourceUrl, workflowMetricRecord, String.format("There's %s workflows in the repository with the same primary descriptor path '%s': %s",
+                    foundFullWorkflowPaths.size(), primaryDescriptorPathFromUrl, foundFullWorkflowPaths), skippedExecutionsCsvPrinter, true);
+             */
+            return Optional.empty();
+        } else {
+            final SourceUrlTrsInfo sourceUrlTrsInfo = new SourceUrlTrsInfo(sourceUrl, "#workflow/" + foundFullWorkflowPaths.get(0), version);
+            return Optional.of(sourceUrlTrsInfo);
+        }
+    }
+
+    /**
+     * Returns a list of slash-delimited components from the source_url.
+     * Example: given source_url https://raw.githubusercontent.com/theiagen/public_health_viral_genomics/v2.0.0/workflows/wf_theiacov_fasta.wdl,
+     * returns a list of "theiagen", "public_health_viral_genomics", "v2.0.0", "wf_theiacov_fasta.wdl"
+     * @param sourceUrl The source_url that starts with https://raw.githubusercontent.com/.
+     * @return
+     */
+    static List<String> getSourceUrlComponents(String sourceUrl) {
+        final String sourceUrlWithoutGitHubPrefix = sourceUrl.replace("https://raw.githubusercontent.com/", "");
+        return Arrays.stream(sourceUrlWithoutGitHubPrefix.split("/"))
+                .filter(urlComponent -> !urlComponent.isEmpty()) // Filter out empty strings that are a result of consecutive slashes '//'
+                .toList();
+    }
+
+    private Optional<String> getPrimaryDescriptorAbsolutePath(WorkflowsApi workflowsApi, MinimalWorkflowInfo workflow, String version) {
+        Optional<String> primaryDescriptorPath = Optional.empty();
+        try {
+            primaryDescriptorPath = Optional.of(workflowsApi.primaryDescriptor1(workflow.id(), version, workflow.descriptorType().toString()).getAbsolutePath());
+        } catch (ApiException e) {
+            LOG.debug("Could not get primary descriptor path", e);
+        }
+        return primaryDescriptorPath;
+    }
+
+    static String makePathAbsolute(String path) {
+        return path.startsWith("/") ? path : "/" + path;
+    }
+
+    /**
+     * Record that stores the calculated TRS ID and version, derived from the source_url
+     * @param sourceUrl
+     * @param trsId
+     * @param version
+     */
+    public record SourceUrlTrsInfo(String sourceUrl, String trsId, String version) {
+    }
+
+    public record MinimalWorkflowInfo(long id, String fullWorkflowPath, DescriptorTypeEnum descriptorType, ConcurrentHashMap<String, String> versionToPrimaryDescriptorPath) {
```

Comment:
Nits, I suggest `ConcurrentMap`, the interface, instead of `ConcurrentHashMap`, even though it's unlikely we would ever change the underlying class. And I also suggest adding `map` to become `versionToPrimaryDescriptorPathMap` -- I was initially confused reading the code further up, e.g., line 510, thinking at first that was an `Optional`.